src/common/blkdev.cc

```diff
@@ -181,82 +148,70 @@ int64_t get_block_device_int_property(const char *devname, const char *property)
   return r;
 }
 
-bool block_device_support_discard(const char *devname)
+bool BlkDev::block_device_support_discard()
 {
-  return get_block_device_int_property(devname, "queue/discard_granularity") > 0;
+  return get_block_device_int_property(BLKDEV_PROP_DISCARD_GRANULARITY) > 0;
 }
 
-int block_device_discard(int fd, int64_t offset, int64_t len)
+int BlkDev::block_device_discard(int64_t offset, int64_t len)
 {
   uint64_t range[2] = {(uint64_t)offset, (uint64_t)len};
   return ioctl(fd, BLKDISCARD, range);
 }
 
-bool block_device_is_rotational(const char *devname)
+bool BlkDev::block_device_is_nvme()
 {
-  return get_block_device_int_property(devname, "queue/rotational") > 0;
+  char vendor[80];
+  // nvme has a device/device/vendor property; infer from that.  There is
+  // probably a better way?
+  int r = get_block_device_string_property(BLKDEV_PROP_VENDOR, vendor, 80);
+  return (r > 0);
 }
 
-int block_device_model(const char *devname, char *model, size_t max)
+bool BlkDev::block_device_is_rotational()
 {
-  return get_block_device_string_property(devname, "device/model", model, max);
+  return get_block_device_int_property(BLKDEV_PROP_ROTATIONAL) > 0;
 }
 
-int get_device_by_uuid(uuid_d dev_uuid, const char* label, char* partition,
-	char* device)
+int BlkDev::block_device_dev(char *dev, size_t max)
 {
-  char uuid_str[UUID_LEN+1];
-  char basename[PATH_MAX];
-  const char* temp_partition_ptr = NULL;
-  blkid_cache cache = NULL;
-  blkid_dev dev = NULL;
-  int rc = 0;
-
-  dev_uuid.print(uuid_str);
-
-  if (blkid_get_cache(&cache, NULL) >= 0)
-    dev = blkid_find_dev_with_tag(cache, label, (const char*)uuid_str);
-  else
-    return -EINVAL;
+  return get_block_device_string_property(BLKDEV_PROP_DEV, dev, max);
+}
 
-  if (dev) {
-    temp_partition_ptr = blkid_dev_devname(dev);
-    strncpy(partition, temp_partition_ptr, PATH_MAX);
-    rc = get_block_device_base(partition, basename,
-      sizeof(basename));
-    if (rc >= 0) {
-      strncpy(device, basename, sizeof(basename));
-      rc = 0;
-    } else {
-      rc = -ENODEV;
-    }
-  } else {
-    rc = -EINVAL;
-  }
+int BlkDev::block_device_model(char *model, size_t max)
+{
+  return get_block_device_string_property(BLKDEV_PROP_MODEL, model, max);
+}
 
-  /* From what I can tell, blkid_put_cache cleans up dev, which
-   * appears to be a pointer into cache, as well */
-  if (cache)
-    blkid_put_cache(cache);
-  return rc;
+int BlkDev::block_device_serial(char *serial, size_t max)
+{
+  return get_block_device_string_property(BLKDEV_PROP_SERIAL, serial, max);
```

Comment:
@asomers Am I right that BLKDEV_PROP_SERIAL property exists only  on NVME devices? Then probably `block_device_serial` would be a little confusing, as someone would expect it returned serial for any device.

Reply:
On Linux I don't have any NVME devices or SAS devices to test with, but SATA and Virt-IO devices lack serial numbers.  On FreeBSD, however, every type of disk does have a serial number.  I think of it more like a Linux bug that SATA serial numbers aren't reported.